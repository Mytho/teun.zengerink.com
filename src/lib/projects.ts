import fs from 'fs'
import path from 'path'
import slugify from 'slugify'
import sizeOf from 'image-size'
import yaml from 'js-yaml'

interface ConfigCaption {
  key: string
  caption: string
}

interface ConfigItem {
  title: string
  statement?: string
  captions?: ConfigCaption[]
}

export interface Photo {
  key: string
  url: string
  size: { width: number; height: number }
  caption?: string
}

export interface Project {
  title: string
  slug: string
  photos: Photo[]
  statement?: string
}

const generatePhotos = (slug: string, captions: ConfigCaption[]): Photo[] => {
  const photosDirectory = path.join(process.cwd(), 'public/photos', slug)
  const fileNames = fs.readdirSync(photosDirectory)

  return fileNames.map((filename) => {
    const key = path.basename(filename, '.jpg')
    const caption = captions?.find((c) => c.key == key)

    return {
      key,
      caption: caption?.caption ?? '',
      url: `/photos/${slug}/${filename}`,
      size: sizeOf(path.join(photosDirectory, filename)),
    }
  })
}

const generateProjects = (item: ConfigItem[]): Project[] => {
  return item.map((project) => {
    const slug = slugify(project.title, { lower: true })

    return {
      slug,
      title: project.title,
      statement: project.statement ?? '',
      photos: generatePhotos(slug, project.captions),
    }
  })
}

export const getProjects = async (): Promise<Project[]> => {
  try {
    const file = path.join(process.cwd(), 'src', 'config.yml')
    const config = yaml.load(fs.readFileSync(file, 'utf-8'))
    return await generateProjects(config)
  } catch (e) {
    console.error(e)
  }
}
