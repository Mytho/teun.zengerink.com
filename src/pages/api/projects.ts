import fs from "fs";
import { IncomingMessage, ServerResponse } from 'http'
import path from "path";
import slugify from "slugify";

interface ConfigCaption {
  key: string;
  caption: string;
}

interface ConfigItem {
  title: string;
  statement?: string;
  captions?: ConfigCaption[];
}

interface Photo {
  key: string;
  url: string;
  caption?: string;
}

export interface Project {
  title: string;
  slug: string;
  photos: Photo[];
  statement?: string;
}

const generatePhotos = (slug: string, captions: ConfigCaption[]): Photo[] => {
  const photosDirectory = path.join(process.cwd(), "public/photos", slug);
  const fileNames = fs.readdirSync(photosDirectory);

  return fileNames.map((filename) => {
    const key = path.basename(filename, ".jpg");
    const caption = captions?.find((c) => c.key == key)?.caption;

    return {
      key,
      caption,
      url: `photos/${slug}/${filename}`,
    };
  });
};

const generateProjects = (item: ConfigItem[]): Project[] => {
  return item.map((project) => {
    const slug = slugify(project.title, { lower: true });

    return {
      title: project.title,
      statement: project.statement,
      slug,
      photos: generatePhotos(slug, project.captions),
    };
  });
};

const config: ConfigItem[] = [
  {
    title: "Take, Make, Waste",
    statement: "TODO",
    captions: [
      {
        key: "01",
        caption: "Image 01: TODO",
      },
      {
        key: "02",
        caption: "Image 02: TODO",
      },
      {
        key: "03",
        caption: "Image 03: TODO",
      },
    ],
  },
  {
    title: "Point of no Return",
    statement: "TODO",
  },
  {
    title: "Under Your Nose",
    statement: "TODO",
  },
  {
    title: "Fortress",
    statement: "TODO",
  },
  {
    title: "Symbiosis",
    statement: "TODO",
  },
  {
    title: "Untitled",
  },
];

const handler = (request: IncomingMessage, response: ServerResponse): void => {
  response.statusCode = 200;
  response.setHeader("Content-Type", "application/json");
  response.end(JSON.stringify({ projects: generateProjects(config) }));
};

export default handler;
