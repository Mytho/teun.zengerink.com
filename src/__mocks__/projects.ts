import { Project } from '../lib/projects'

const projects: Project[] = [
  {
    title: 'My First Project',
    slug: 'my-first-project',
    statement: 'The statement for the project',
    photos: [
      {
        key: '01',
        url: '/photos/my-first-project/01.jpg',
        size: { width: 800, height: 600 },
      },
      {
        key: '02',
        url: '/photos/my-first-project/02.jpg',
        size: { width: 600, height: 800 },
      },
      {
        key: '03',
        url: '/photos/my-first-project/03.jpg',
        size: { width: 800, height: 600 },
      },
    ],
  },
  {
    title: 'Project without a statement',
    slug: 'no-statement',
    photos: [
      {
        key: '01',
        url: '/photos/no-statement/01.jpg',
        caption: 'This is image 1',
        size: { width: 800, height: 600 },
      },
      {
        key: '02',
        url: '/photos/no-statement/02.jpg',
        caption: 'This is image 2',
        size: { width: 800, height: 600 },
      },
      {
        key: '03',
        url: '/photos/no-statement/03.jpg',
        caption: 'This is image 3',
        size: { width: 600, height: 800 },
      },
      {
        key: '04',
        url: '/photos/no-statement/04.jpg',
        caption: 'This is image 4',
        size: { width: 800, height: 600 },
      },
    ],
  },
]

export default projects
