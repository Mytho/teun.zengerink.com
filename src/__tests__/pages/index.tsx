import React from 'react'
import renderer from 'react-test-renderer'
import Home from '../../pages/index'
import projects from '../../__mocks__/projects'

const defaultProps = { projects }

it('renders correctly', () => {
  const tree = renderer.create(<Home {...defaultProps} />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('displays a loader when no projects are loaded', () => {
  const props = { projects: [] }
  const tree = renderer.create(<Home {...props} />).toJSON()
  expect(tree).toMatchSnapshot()
})
