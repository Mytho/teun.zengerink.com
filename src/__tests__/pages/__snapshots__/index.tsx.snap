// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays a loader when no projects are loaded 1`] = `
<div>
  <aside
    className="sidebar "
  >
    <a
      href="/"
    >
      <h1>
        Teun Zengerink
      </h1>
    </a>
    <div
      className="hamburger"
      onClick={[Function]}
    >
      <span
        className="line"
      />
      <span
        className="line"
      />
      <span
        className="line"
      />
    </div>
    <nav
      className="navigation"
    >
      
      <ul />
      <span
        className="socials"
      >
        <a
          href="https://instagram.com/t.zengerink/"
        >
          <img
            alt="Instagram"
            src="/img/instagram.jpg"
          />
        </a>
        <a
          href="https://pinterest.com/tzengerink/"
        >
          <img
            alt="Pinterest"
            src="/img/pinterest.jpg"
          />
        </a>
      </span>
    </nav>
  </aside>
  <main
    className="main"
  >
    <div
      className="loader"
    />
  </main>
</div>
`;

exports[`renders correctly 1`] = `
<div>
  <aside
    className="sidebar "
  >
    <a
      href="/"
    >
      <h1>
        Teun Zengerink
      </h1>
    </a>
    <div
      className="hamburger"
      onClick={[Function]}
    >
      <span
        className="line"
      />
      <span
        className="line"
      />
      <span
        className="line"
      />
    </div>
    <nav
      className="navigation"
    >
      <h2>
        Work
      </h2>
      <ul>
        <li
          className=""
        >
          <a
            href="/work/my-first-project"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            My First Project
          </a>
        </li>
        <li
          className=""
        >
          <a
            href="/work/no-statement/01"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Project without a statement
          </a>
        </li>
      </ul>
      <span
        className="socials"
      >
        <a
          href="https://instagram.com/t.zengerink/"
        >
          <img
            alt="Instagram"
            src="/img/instagram.jpg"
          />
        </a>
        <a
          href="https://pinterest.com/tzengerink/"
        >
          <img
            alt="Pinterest"
            src="/img/pinterest.jpg"
          />
        </a>
      </span>
    </nav>
  </aside>
  <main
    className="main"
  >
    <div
      className="container"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <div
        className="title"
      >
        My First Project
      </div>
      <img
        alt="My First Project"
        height={600}
        src="/photos/my-first-project/01.jpg"
        width={800}
      />
    </div>
  </main>
</div>
`;
